protocol/igmp.go

```diff
@@ -0,0 +1,421 @@
+}
+
+func NewIGMPv1Query(group net.IP) *IGMPv1or2 {
+	return &IGMPv1or2{Type: IGMPQuery, GroupAddress: group}
+}
+
+func NewIGMPv1Report(group net.IP) *IGMPv1or2 {
+	return &IGMPv1or2{Type: IGMPv1Report, GroupAddress: group}
+}
+
+func NewIGMPv2Query(group net.IP, maxResponseTime uint8) *IGMPv1or2 {
+	return &IGMPv1or2{
+		Type:            IGMPQuery,
+		MaxResponseTime: maxResponseTime,
+		GroupAddress:    group,
+	}
+}
+
+func NewIGMPv2Report(group net.IP) *IGMPv1or2 {
+	return &IGMPv1or2{
+		Type:         IGMPv2Report,
+		GroupAddress: group,
+	}
+}
+
+func NewIGMPv2Leave(group net.IP) *IGMPv1or2 {
+	return &IGMPv1or2{
+		Type:         IGMPv2LeaveGroup,
+		GroupAddress: group,
+	}
+}
+
+// IGMPv3Query:
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |  Type = 0x11  | Max Resp Code |           Checksum            |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                         Group Address                         |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   | Resv  |S| QRV |     QQIC      |     Number of Sources (N)     |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                       Source Address [1]                      |
+//   +-                                                             -+
+//   |                       Source Address [2]                      |
+//   +-                              .                              -+
+//   .                               .                               .
+//   .                               .                               .
+//   +-                                                             -+
+//   |                       Source Address [N]                      |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IGMPv3Query struct {
+	Type                     uint8
+	MaxResponseTime          uint8
+	Checksum                 uint16
+	GroupAddress             net.IP
+	Reserved                 uint8
+	SuppressRouterProcessing bool
+	RobustnessValue          uint8
+	IntervalTime             uint8
+	NumberOfSources          uint16
+	SourceAddresses          []net.IP
+}
+
+func (p *IGMPv3Query) Len() uint16 {
+	return 12 + p.NumberOfSources*4
+}
+
+func (p *IGMPv3Query) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(p.Len()))
+	n := 0
+	data[n] = p.Type
+	n += 1
+	data[n] = p.MaxResponseTime
+	n += 1
+	binary.BigEndian.PutUint16(data[n:], p.Checksum)
+	n += 2
+	copy(data[n:n+4], p.GroupAddress[12:16])
+	n += 4
+	sBit := uint8(0x0)
+	if p.SuppressRouterProcessing {
+		sBit = 0x8
+	}
+	data[n] = sBit | p.RobustnessValue&0x7
+	n += 1
+	data[n] = p.IntervalTime
+	n += 1
+	binary.BigEndian.PutUint16(data[n:], p.NumberOfSources)
+	n += 2
+	if p.NumberOfSources > 0 {
```

Comment:
Nit: It would be cleaner to leave out the 'if' statement. 'range' will any how not run if there are no elements.